engine/utils/collections/Iterators.h

```diff
@@ -99,6 +99,92 @@ class Iter
     A* ptr {nullptr};
 };
 
+/**
+ * A standard linear iterator. This iterator always iterates over every element of the array
+ * regardless of whether it has been filled or not. This is the 'standard' iterator type
+ * @tparam T the type of array stored by the iterator. Must be an array type which exposes a
+ * 'Data()' and `Size()` function
+ * @tparam R The type returned by the iterator (usually the type of the array)
+ */
+template<typename A, typename T>
+class CIter
+{
+public:
+
+    typedef void (CIter::*BoolType)() const;
+
+    // TODO(Aryeh): Add more operators as needed (--, ->, etc).
+    /**
+     * @brief CreateIterator constructor
+     * @param arrPtr the pointer to the Array
+     */
+    inline constexpr CIter(const A* arrPtr)
+    {
+        ptr = arrPtr->Data() && arrPtr->Size() > 0 ? arrPtr : nullptr;
+    }
+
+    /**
+     * @brief The iterator prefix increment operator. This operator will increment the pointer
+     * as many times as needed to find an active element. As such, it will not return inactive
+     * array elements
+     * @return the iterator with the pointer and index incremented
+     */
+    inline constexpr CIter& operator++()
+    {
+        // TODO: Profile if using a reinterpret cast would be faster here
+        ptr = ++index < ptr->Size() ? ptr : 0;
+        return *this;
+    }
+
+    /**
+     * @brief The dereference operator
+     * @return the de-referenced pointer value
+     */
+    inline constexpr const T& operator*() const
+    {
+        return *(ptr->Data() + index);
+    }
+
+    /**
+     * Checks if the CreateIterator if valid or not
+     * @return true of the CreateIterator is valid, false otherwise
+     */
+    inline constexpr operator BoolType() const
+    {
+        return ptr ? &CIter::DoNothing : 0;
```

Comment:
`return (bool) ptr`?

Reply:
Not possible. This wouldn't work in the framework of how the boolean conversion idiom works.